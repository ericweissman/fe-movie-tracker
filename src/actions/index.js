export const getMovies = (movies) => ({
  type: 'GET_MOVIES',
  movies
});

export const loginUser = (user) => ({
  type: 'LOGIN_USER',
  user
})

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
})

export const populateFavorites = (favorites) => ({
  type: 'POPULATE_FAVORITES',
  favorites
})
