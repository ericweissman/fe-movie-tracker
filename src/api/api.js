export const fetchData = async (url) => {
  const response = await fetch(url);
  if(response.ok) {
    return response.json()
  } else {
    throw Error(`Error fetching, ${response.status}`)
  }
}

export const postData = async (urlSuffix, data) => {
  const url = 'http://localhost:3000/api/users' + urlSuffix;
  const response = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'application/json'
    }
  })
  if(response.ok) {
    let data = response.json();
    return data
  } else {
    throw Error(`Error fetching, ${response.status}`)
  }
}