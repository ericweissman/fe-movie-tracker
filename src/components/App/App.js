import React, { Component } from 'react';
import './App.scss';
import { apiKey } from '../../utils/apiKey'
import { fetchData } from '../../api/api'
import { connect } from 'react-redux'
import { getMovies } from '../../actions'
import { NavLink, Route } from 'react-router-dom'
import { withRouter } from 'react-router'
import  MovieArea  from '../MovieArea/MovieArea'
import Login from '../Login/Login'
import CreateUser from '../CreateUser/CreateUser'



class App extends Component {
  constructor() {
    super()
  }

  componentDidMount = async () => {
    const url = `https://api.themoviedb.org/3/movie/popular?api_key=${apiKey}&language=en-US&page=1`;
    const movies = await fetchData(url)
    //add cleaner function?
    this.props.getMovies(movies.results)
  }

  render() {
    return (
      <div className="App">
        <header>
          <NavLink to='/' className='nav'>Show All Movies</NavLink>
          <NavLink to='/login' className='nav'>Login</NavLink>
          <NavLink to='/signup' className='nav'>Signup</NavLink>
          <NavLink to='/favorites' className='nav' >View Favorites</NavLink>
        </header>
        <Route exact path='/' component={MovieArea}/>
        <Route path='/login' component={Login}/>
        <Route path='/signup' component={CreateUser}/>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  movies: state.movies
})

const mapDispatchToProps = (dispatch) => ({
  getMovies: (movies) => dispatch(getMovies(movies))
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));
