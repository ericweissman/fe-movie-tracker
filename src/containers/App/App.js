import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Route } from 'react-router-dom'
import { withRouter } from 'react-router'
import './App.scss'
import '../../index.scss'
import { apiKey } from '../../utils/apiKey'
import { fetchData } from '../../api/api'
import { getMovies, logoutUser } from '../../actions'
import { Header } from '../../components/Header/Header'
import MovieArea from '../MovieArea/MovieArea'
import MovieDetails from '../../components/MovieDetails/MovieDetails'
import Login from '../../containers/Login/Login'
import CreateUser from '../CreateUser/CreateUser'

export class App extends Component {

  componentDidMount = async () => {
    const url = `https://api.themoviedb.org/3/movie/popular?api_key=${apiKey}&language=en-US&page=1`
    let movies = await fetchData(url)
    movies = movies.results.map((movie) => {
      return {
        movie_id: movie.id,
        title: movie.title,
        poster_path: movie.poster_path,
        release_date: movie.release_date,
        vote_average: movie.vote_average,
        overview: movie.overview
      }
    })
    this.props.getMovies(movies)
  }

  render() {
    const { logoutUser, movies, user } = this.props
    return (
      <div className='App'>
        <Header logoutUser={logoutUser} user={user} />
        <Route exact path='/' component={MovieArea} />
        <Route exact path='/login' component={Login} />
        <Route exact path='/signup' component={CreateUser} />
        <Route exact path='/favorites' component={MovieArea} />
        <Route
          path='/movies/:id'
          render={({ match }) => {
            const { id } = match.params
            const movie = movies.find((movie) => {
              return movie.movie_id === parseInt(id)
            })
            if (movie) {
              return <MovieDetails {...movie} />
            }
          }}
        />
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  movies: state.movies,
  user: state.user
})

export const mapDispatchToProps = (dispatch) => ({
  getMovies: (movies) => dispatch(getMovies(movies)),
  logoutUser: () => dispatch(logoutUser())
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App))
