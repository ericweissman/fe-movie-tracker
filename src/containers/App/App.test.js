import React from 'react';
import ReactDOM from 'react-dom';
import {App, mapDispatchToProps, mapStateToProps } from './App';
import { shallow } from 'enzyme';
import * as api from '../../api/api';
import { getMovies, loginUser } from '../../actions'
describe('App', () => {
let wrapper;
const mockUser = {name: 'eric', id: 2, favorites: [2]}
const mockGetMovies = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<App user={mockUser} getMovies={mockGetMovies}/>)
  });

  it('should match the correct snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    it('should call fetchData the correct number of times', () => {
      api.fetchData = jest.fn(() => {results: [{movie_id: 2, title: 'Birds', poster_path: 'sfsafas', release_date: '1990-01-01', vote_average: 4.4, overview: 'assafasfas'}]});
      wrapper.instance().componentDidMount();
      expect(api.fetchData).toHaveBeenCalled();
    });

    it('should call getMovies with the correct params', () => {
      api.fetchData = jest.fn({page: 1, total_results: 19845, total_pages: 993, results: [{movie_id: 2, title: 'Birds', poster_path: 'sfsafas', release_date: '1990-01-01', vote_average: 4.4, overview: 'assafasfas'}]})
      wrapper.instance().componentDidMount();
      expect(wrapper.instance().props.getMovies).toHaveBeenCalled();
      // expect(mockGetMovies).toHaveBeenCalled();
    })
  })

  describe('mapStateToProps', () => {
    it('should return an object with the movies array', () => {
      const mockState = {
        user: {name: 'Eric', id: 2, favorites: [3]}
      };
      const expected = {user: {name: 'Eric', id: 2, favorites: [3]}};

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch when using a function from mapDispatchToProps', () => {
      const mockDispatch = jest.fn();
      const movies = { 
        movie_id: 2,
        title: 'Dumb and Dumber',
        poster_path: 'ansfkjasnfjsan.com',
        release_date: '2018-01-01',
        vote_average: 4.1,
        overview: 'Two friends go on a roadtrip.'
       };
      const actionToDispatch = getMovies(movies);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.getMovies(movies);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  } )
})