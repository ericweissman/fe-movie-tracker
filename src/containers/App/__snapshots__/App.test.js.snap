// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the correct snapshot 1`] = `
<div
  className="App"
>
  <Header
    user={
      Object {
        "favorites": Array [
          2,
        ],
        "id": 2,
        "name": "eric",
      }
    }
  />
  <Route
    component={[Function]}
    exact={true}
    path="/"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/login"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/signup"
  />
  <Route
    component={[Function]}
    exact={true}
    path="/favorites"
  />
  <Route
    path="/movies/:id"
    render={[Function]}
  />
</div>
`;
